### PR TITLE
feat: support lsp code actions

### PR DESCRIPTION
**Pull request by sockmaster27**
Adds the framework for supporting code actions. The actions themselves can now be implemented in `CodeActionProvider.scala`.
I can see that the auto-formatting went a bit wild on some parts. I hope that isn't a problem.

Related to flix/vscode-flix#233

### CODE REVIEW

**Comment by magnus-madsen** [comment]
>  I can see that the auto-formatting went a bit wild on some parts. I hope that isn't a problem.

Unfortunately that is a bit of a problem because it makes code review difficult, and more importantly, it makes it harder to browse the git history, e.g. with git blame/bisect etc. I would better if you can go back and change the right IDE setting and then only have actual changes :)

**Comment by magnus-madsen** [comment]
It may or may not not be preferable to have two PRs: one that simply introduced the new data types and one that contains changes to existing ones.

**Comment by sockmaster27** [comment]
> > I can see that the auto-formatting went a bit wild on some parts. I hope that isn't a problem.
> 
> Unfortunately that is a bit of a problem because it makes code review difficult, and more importantly, it makes it harder to browse the git history, e.g. with git blame/bisect etc. I would better if you can go back and change the right IDE setting and then only have actual changes :)

This is caused by the fact that there is some inconsistent formatting in the doc comments. 
Some look like this:
```
/**
 * Doc comment
 */
```
and others look like this:
```
/**
  * Doc comment
  */
```

This is controlled by the "Add additional space for leading asterisk" option in IntelliJ, and as far as I can tell, there is no way to disable that (you have to choose whether they should be there or not). This seems to make it impossible to, for example, format-on-save without messing it up like I've done here.

Could I instead make a different PR to make this consistent across the code base?

**Comment by JonathanStarup** [comment]
I believe this is the "correct" one.
```
/**
  * Doc comment
  */
```
I've thought about fixing it project-wide for a while so I support the idea 😄 but it looks like if you just change your setting then most of the noise should go away

**Comment by magnus-madsen** [comment]
While there is (apparently) some whitespace issue, we should resolve that separately from this PR.

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:51]: extra `{}`

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:29]: 1. I'm surprised we have to roll our own parser here. Does `json4s` not provide conversion-to-case-class facilities?
2. If not, I think we can abstract away some of the common functionality here, like `getOptionalString` and `getOptionalList` seem to be generally useful.

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:1]: License

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:12]: May be slightly misleading if we see e.g. `not recognized as a valid integer: '3'.`

File: main/src/ca/uwaterloo/flix/api/lsp/DiagnosticSeverity.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/DiagnosticSeverity.scala:48]: see above

File: main/src/ca/uwaterloo/flix/api/lsp/Diagnostic.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/Diagnostic.scala:53]: See above regarding parsing

File: main/src/ca/uwaterloo/flix/api/lsp/DiagnosticTag.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/DiagnosticTag.scala:35]: see above regarding integer

File: main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala:1]: license

**Review by mlutze — CHANGES_REQUESTED** [review]

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:29]: We usually do roll our own parsers. My surprise is that w eneed to parse diagnostics.

But I think @sockmaster27 is just faithfully implementing the API.

But I suspect we can get away with something simpler.

Q: CodeActionContext is something sent _to us_ right?

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionKind.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/CodeActionKind.scala:24]: Empty string or undefined?

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:29]: Yes

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:29]: If codeactioncontext is not sent to us, why do we need to parse it?

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:20]: I think it would be better with two separate error messages.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:20]: But more to the point, according to the protocol these unknown values should not really occur. Hence it might be better for us to throw exceptions?

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:80]: I still see some unnecessary changes to this file. I think it might be best to checkout the original file and then port your changes to it.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:80]: The reason I care is because I don't want to break "git blame" etc.

**Comment by magnus-madsen** [comment]
This is looking really nice 👍 

Matt and I just have several comments with us trying to understand the code.

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:29]: It _is_ sent to us

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionKind.scala
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionKind.scala:24]: Empty string: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeActionKind

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:20]: I think it's more in line with the rest of the code to treat it as a `Result`

File: main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:80]: Sorry, I somehow missed these 🤔

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:20]: In general you are totally write. But in this case its really should not fail because if it does something is very wrong.

In any case, I suspect that if we remove the list of diagnostics then we can remove this parsing code, and it becomes moot.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionTriggerKind.scala:20]: right*

File: main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:25]: Removing parsing of `Diagnostic` pretty much eliminates this, so unless there are any objections I'll mark this as resolved

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/CodeActionContext.scala:25]: Sounds good 👍

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@mlutze Can you unblock this PR?

**Comment by magnus-madsen** [comment]
@sockmaster27 I tried commenting in the code in `CodeActionProvider`, but I don't see any code actions?

**Comment by sockmaster27** [comment]
@magnus-madsen 
Hmm that's weird. I tried that just yesterday. Are you sure you're using the right version of the plugin and loading the right version of the compiler?

**Comment by magnus-madsen** [comment]
I thought so. I grabbed both masters after merging, build myself a jar, and am running the vscode extension.

On the other hand there are no errors either-- which would usually be the case.

**Comment by magnus-madsen** [comment]
Wait! I see it. The cursor just has to be at the right position.

**Comment by magnus-madsen** [comment]
I guess it only shows depending on the cursor or?

**Comment by sockmaster27** [comment]
@magnus-madsen
It should show up as a little light bulb next to the line that the cursor is on (or sometimes just above that line). It might not show up if the line is empty? I don't remember, but if you're seeing it I guess it should be fine, right?

**Comment by magnus-madsen** [comment]
Yes. At least we can figure it out once we start to add actual fixes.

**Pull request merged**
